### PR TITLE
【KernelGen】Add bincount operator

### PR DESCRIPTION
**Pull request by Schopenhauer-loves-Hegel**
### PR Category
Operator

### Type of Change
New Feature

### Description
Add `bincount` operator implementation with Triton kernel.

- Implementation mode: `N/A`
- Accuracy test: 90/90 passed

### Issue
N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance
**torch.int64**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| torch.Size([1000]) | 0.0840 | 0.1408 | 0.597 |
| torch.Size([10000]) | 0.0886 | 0.1400 | 0.633 |
| torch.Size([10000]) | 0.0835 | 0.1401 | 0.596 |
| torch.Size([100000]) | 0.1250 | 0.1677 | 0.745 |
| torch.Size([100000]) | 0.1277 | 0.1417 | 0.901 |
| torch.Size([1000000]) | 0.1087 | 0.5714 | 0.190 |
| torch.Size([1000000]) | 5.0190 | 0.2536 | 19.794 |
| torch.Size([1000000]) | 0.1237 | 0.1618 | 0.764 |

**Overall: median speedup = 0.689x, mean speedup = 3.027x** (8 data points)

---
_Generated by auto_gen tool with Claude Code_

### CODE REVIEW

File: benchmark/test_special_perf.py [hunk 1]
```diff
@@ -245,18 +216,9 @@ def resolve_conj_input_fn(shape, dtype, device):
     yield x.conj(),
 
 
-@pytest.mark.topk
```
Comment (tengqm) [inline benchmark/test_special_perf.py:248]: why remove this test case?

File: benchmark/test_special_perf.py [hunk 2]
```diff
@@ -1521,6 +1483,87 @@ def test_functional_sym_constrain_range_for_size():
     bench.run()
 
 
+class BincountBenchmark(Benchmark):
+    """Benchmark for bincount operation."""
+
+    def __init__(self, op_name, torch_op, dtypes):
+        super().__init__(op_name=op_name, torch_op=torch_op, dtypes=dtypes)
+
```
Comment (tengqm) [inline benchmark/test_special_perf.py:1544]: is this needed?

File: benchmark/test_special_perf.py [hunk 3]
```diff
@@ -1521,6 +1483,87 @@ def test_functional_sym_constrain_range_for_size():
...
+        inp = torch.randint(0, max_val, (input_size,), dtype=torch.int64, device=device)
+        yield inp,
+
+
+class BincountWeightsBenchmark(Benchmark):
+    """Benchmark for bincount operation with weights."""
+
+    def __init__(self, op_name, torch_op, dtypes):
+        super().__init__(op_name=op_name, torch_op=torch_op, dtypes=dtypes)
+
```
Comment (tengqm) [inline benchmark/test_special_perf.py:1573]: this function is not needed?

File: src/flag_gems/ops/__init__.py [hunk 4]
```diff
@@ -394,6 +395,7 @@
     "bitwise_or_tensor",
     "bitwise_or_tensor_",
     "bitwise_right_shift",
+    "bincount",
```
Comment (tengqm) [inline src/flag_gems/ops/__init__.py:398]: not sorted again

File: src/flag_gems/ops/bincount.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
+import logging
+
+import torch
+import triton
+import triton.language as tl
+
+from ..utils import libentry
```
Comment (tengqm) [inline src/flag_gems/ops/bincount.py]:
```suggestion
from flag_gems.utils import libentry
```

File: src/flag_gems/ops/bincount.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        # Output dtype matches weights dtype
+        # For atomic_add compatibility, convert to float32 if float16/bfloat16
+        weights_dtype = weights.dtype
+        if weights_dtype in (torch.float16, torch.bfloat16):
+            weights = weights.to(torch.float32)
+            out = torch.zeros(output_size, dtype=torch.float32, device=inp.device)
+        else:
+            out = torch.zeros(output_size, dtype=weights.dtype, device=inp.device)
+
+        BLOCK_SIZE = 1024
```
Comment (tengqm) [inline src/flag_gems/ops/bincount.py:108]: not very sure if hardcoding BLOCK_SIZE is good for all backends.

Comment (tengqm) [reply src/flag_gems/ops/bincount.py:108]: But I'm okay with this implementation at the moment.

**Review by tengqm — COMMENTED** [review]
I don't know why this PR is removing a lot of existing test cases.

**Comment by factnn** [comment]
## Response to Review Comments

**Q1: Lines 1543, 1572 - Are `super().__init__()` calls needed?**

Yes, they are required. The `Benchmark` base class `__init__` method (defined in `benchmark/performance_utils.py:83-101`) initializes essential attributes including `self.op_name`, `self.torch_op`, `self.gems_op`, `self._input_iter`, and dtype/shape configurations. Without calling `super().__init__()`, `bench.run()` will fail due to missing attributes.

**Q2: Line 108 - Is hardcoding `BLOCK_SIZE=1024` good for all backends?**

Good point. The hardcoded `BLOCK_SIZE=1024` is optimized for NVIDIA GPUs but may not be optimal for other backends (e.g., AMD with wavefront size 64). TODO: Consider using `@triton.autotune` or backend-specific block size selection for better cross-platform performance. This can be addressed as a follow-up optimization.

File: benchmark/test_special_perf.py (hunk 2 above)
Comment (tengqm) [reply benchmark/test_special_perf.py]: I'm not convinced that this is needed because the method prototype is exactly the same. Am I missing something?

File: benchmark/test_special_perf.py [hunk 7]
```diff
@@ -1535,3 +1535,81 @@ def unique_consecutive_input_fn(shape, dtype, device):
...
+    def bincount_weights_input_fn(self, config, dtype, device):
+        input_size, max_val = config
+        inp = torch.randint(0, max_val, (input_size,), dtype=torch.int64, device=device)
+        weights = torch.randn(input_size, dtype=dtype, device=device)
+        yield inp, {"weights": weights}
+
+
+@pytest.mark.bincount
+def test_perf_bincount():
```
Comment (tengqm) [inline benchmark/test_special_perf.py]: `perf` is wordy and redundant ... please avoid using it.

```suggestion
def test_bincount():
```

File: benchmark/test_special_perf.py [hunk 8]
```diff
@@ -1535,3 +1535,81 @@ def unique_consecutive_input_fn(shape, dtype, device):
...
+    bench = BincountBenchmark(
+        op_name="bincount",
+        torch_op=torch.bincount,
+        dtypes=[torch.int64],
+    )
+    bench.run()
+
+
+@pytest.mark.bincount
+def test_perf_bincount_with_weights():
```
Comment (tengqm) [inline benchmark/test_special_perf.py]:
```suggestion
def test_bincount_with_weights():
```

File: benchmark/test_special_perf.py [hunk 9]
```diff
@@ -1535,3 +1535,81 @@ def unique_consecutive_input_fn(shape, dtype, device):
         dtypes=INT_DTYPES,
     )
     bench.run()
+
+
+class BincountBenchmark(Benchmark):
```
Comment (tengqm) [inline benchmark/test_special_perf.py]: You may want to put this into a separate benchmark file as well, just like we have separate accuracy test case file now.

**Comment by factnn** [comment]
## UT Results

Branch rebased to upstream/master and tests passed:

```
python -m pytest tests/ -m bincount --ref cpu -x -q
90 passed
```

**Review by w1120029931-bit — APPROVED** [review]
